### ISSUE TITLE
Lint for likely-malformed reference style links

### ISSUE THREAD

**Issue by CAD97** [rust-lang-nursery/rust-clippy#1952]
I often come across the situation in markdown when I accidentally write the following:

```markdown
[useful text][reference]

[reference](<example.com>)
```

This, though it might look like a reference style link, is not:

> [useful text][reference]
> 
> [reference](<example.com>)

The correct syntax is as follows:

```markdown
[useful text][reference]

[reference]: <example.com>
```

> [useful text][reference1]
> 
> [reference1]: <example.com>

Clippy could lint against this likely-malformed reference link.

Notes:

- This lint should only trigger if `[reference](<link>)` is on a line on its own (e.g. would be a valid reference target) and `reference` could be a valid reference key.
- This lint should not trigger if there is any escaping of the in-text part of the reference which makes it not a valid reference link, even if it renders \[like]\[so] in the output.
- I don't think \[bracketed]\[words] would ever be desired except in a code block about markdown, so this lint shouldn't have a lot of real-world false positives.

**Comment by CAD97** [comment]
Nevermind I had the reference style link syntax wrong and fixing it made clippy ignore it like it should

**Comment by llogiq** [comment]
We could still try to detect those cases and issue a better warning message.

**Comment by CAD97** [comment]
What I had was `[ref](link)` instead of `[ref]: link` if you want to lint against it @llogiq. It feels like a common mistake (at least for me).

**Comment by CAD97** [comment]
I have re-purposed this issue to lint against near-miss reference style links.
